pkg/ruler/manager.go

```diff
@@ -353,6 +359,27 @@ func (r *DefaultMultiTenantManager) removeUsersIf(shouldRemove func(userID strin
 		level.Info(r.logger).Log("msg", "deleted rule manager and local rule files", "user", userID)
 	}
 
+	r.userManagerMtx.Unlock()
+
+	wg.Wait()
+
+	// Now that the rule managers are stopped, let's make sure we also stop the notifiers.
+
+	r.notifiersMtx.Lock()
+	for userID, n := range r.notifiers {
+		if shouldRemove(userID) {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				n.stop()
```

Comment:
Note that once https://github.com/prometheus/prometheus/pull/14290 is merged, this could take a long time, which would block further resharding. We'd need to run this in a way that doesn't block further resharding.